cpp/cmake_modules/ThirdpartyToolchain.cmake

```diff
@@ -217,6 +217,9 @@ endif()
 
 if(ARROW_FLIGHT OR ARROW_IPC)
   set(ARROW_WITH_FLATBUFFERS ON)
+endif()
+
+if(ARROW_JSON OR ARROW_FLIGHT OR ARROW_IPC)
```

Comment:
Now we should only have a single `if(ARROW_JSON)` here, right?
Also, json shouldn't be needed for Flight.

Reply:
> Now we should only have a single `if(ARROW_JSON)` here, right?

whoops, yes, I'll fix this.

> Also, json shouldn't be needed for Flight.

Do you know if Flight requires IPC?  I tried:

```
% cmake ..  -DARROW_IPC=OFF -DARROW_FLIGHT=ON
...
CMake Error at cmake_modules/BuildUtils.cmake:189 (add_dependencies):
  The dependency target "metadata_fbs" of target "arrow_flight_objlib" does
  not exist.
Call Stack (most recent call first):
  src/arrow/flight/CMakeLists.txt:90 (add_arrow_lib)
```

I'm just going to try this out on master to work out if this is something I broke.  Let me know if there is anything else I should change and thanks for your patience on this one.

Reply:
Ah, right, Flight requires IPC. My bad.

Reply:
I think the error I ran into above comes from [cpp/src/arrow/ipc/CMakeLists.txt#L78](https://github.com/apache/arrow/blob/b8cadbbfce2f01095c5bbd6b816de70d36bf071f/cpp/src/arrow/ipc/CMakeLists.txt#L78) which means that flight needs the IPC component.

Reply:
Ok I attempted to resolve this but making sure that IPC is built when Flight is built.  Let me know if this is a step too far?

Reply:
That sounds reasonable. @lihalite what do you think?

Reply:
Yes, sounds good to me.